### PR TITLE
dotnet receiver: process block messages

### PR DESCRIPTION
**Pull request by pmcollins**
This is the fourth PR in a series for the dotnet core diagnostics receiver.

This PR processes messages in a loop after the initial request to and initial responses from the dotnet diagnostic process.

For the full implementation please see #2200

### CODE REVIEW

File: receiver/dotnetdiagnosticsreceiver/dotnet/event_parser.go [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+package dotnet
+
+import (
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/dotnetdiagnosticsreceiver/network"
+)
+
+// parseEventBlock parses an event block and returns a Metric slice containing
+// the raw representation of the metrics extracted from the event messages.
+// All of the previous parsing is for correct operation of the functionality
```
Comment (tigrannajaryan) [inline receiver/dotnetdiagnosticsreceiver/dotnet/event_parser.go:23]: What "previous parsing" does this refer to?

File: receiver/dotnetdiagnosticsreceiver/dotnet/metadata_parser.go [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+	}
+	return
+}
+
+const (
+	typeCodeObject = 1
+	typeCodeInt32  = 9
+	typeCodeSingle = 13
+	typeCodeDouble = 14
+	typeCodeString = 18
```
Comment (tigrannajaryan) [inline receiver/dotnetdiagnosticsreceiver/dotnet/metadata_parser.go:186]: Link to docs please.

File: receiver/dotnetdiagnosticsreceiver/dotnet/parser.go [hunk 3]
```diff
@@ -44,3 +46,71 @@ func (p *Parser) ParseIPC() error {
...
+	if err != nil {
+		return err
+	}
+
+	st, err := parseSerializationType(p.r)
+	if err != nil {
+		return err
+	}
+
+	p.logger.Info("parsing block", zap.String("serialization type", st.name))
```
Comment (tigrannajaryan) [inline receiver/dotnetdiagnosticsreceiver/dotnet/parser.go:81]: Does this need to be `Info`? Would it not flood the logs?

File: receiver/dotnetdiagnosticsreceiver/dotnet/stack_parser.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+package dotnet
+
+import (
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/dotnetdiagnosticsreceiver/network"
+)
+
+// parseStackBlock parses a stack block. The results are currently not used, but
+// skippping over the correct number of bytes is required for parsing subsequent
+// messages.
```
Comment (tigrannajaryan) [inline receiver/dotnetdiagnosticsreceiver/dotnet/stack_parser.go:23]: 👍

File: receiver/dotnetdiagnosticsreceiver/dotnet/trace_parser.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+		pointerSize             int32
+		processID               int32
+		numProcessors           int32
+		expectedCPUSamplingRate int32
+	}
+
+	es := eventSource{}
+
+	// skip date
+	err = w.Seek(16)
```
Comment (tigrannajaryan) [inline receiver/dotnetdiagnosticsreceiver/dotnet/trace_parser.go:38]: Can we define 16 as const?

File: receiver/dotnetdiagnosticsreceiver/receiver.go [hunk 6]
```diff
@@ -78,9 +79,19 @@ func (r *receiver) Start(ctx context.Context, host component.Host) error {
...
+			r.logger.Error("parseAll error", zap.Error(err))
+		}
+	}()
 	return nil
 }
 
 func (r *receiver) Shutdown(context.Context) error {
+	if r.cancel != nil {
+		r.cancel()
+	}
```
Comment (tigrannajaryan) [inline receiver/dotnetdiagnosticsreceiver/receiver.go:95]: 👍

File: receiver/dotnetdiagnosticsreceiver/dotnet/event_header.go [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+type headerFlags byte
+
+const (
+	headerFlagMetadataID               headerFlags = 1 << 0
+	headerFlagCaptureThreadAndSequence headerFlags = 1 << 1
+	headerFlagThreadID                 headerFlags = 1 << 2
+	headerFlagStackID                  headerFlags = 1 << 3
+	headerFlagActivityID               headerFlags = 1 << 4
+	headerFlagRelatedActivityID        headerFlags = 1 << 5
+	headerFlagDataLength               headerFlags = 1 << 7
```
[the remark this answers is not in the document]
Comment (pmcollins) [reply receiver/dotnetdiagnosticsreceiver/dotnet/event_header.go:41]: Sure. I've added a link to the C# file where I got this.

File: receiver/dotnetdiagnosticsreceiver/dotnet/event_parser.go (hunk 1 above)
Comment (pmcollins) [reply receiver/dotnetdiagnosticsreceiver/dotnet/event_parser.go]: Mostly `parseMetadataBlock`. I've updated the comment.

File: receiver/dotnetdiagnosticsreceiver/dotnet/metadata_parser.go (hunk 2 above)
Comment (pmcollins) [reply receiver/dotnetdiagnosticsreceiver/dotnet/metadata_parser.go:187]: Sure, done.

File: receiver/dotnetdiagnosticsreceiver/dotnet/parser.go (hunk 3 above)
Comment (pmcollins) [reply receiver/dotnetdiagnosticsreceiver/dotnet/parser.go]: It's max 1 log message per second, but the information is probably not terribly useful except for debugging. Changed to `Debug`.

File: receiver/dotnetdiagnosticsreceiver/dotnet/trace_parser.go (hunk 5 above)
Comment (pmcollins) [reply receiver/dotnetdiagnosticsreceiver/dotnet/trace_parser.go]: Sure, done.

**Review by tigrannajaryan — APPROVED** [review]

**Pull request merged**